File: app/src/main/java/com/bhupendra/foodorderingapp/MainActivity.java
package com.bhupendra.foodorderingapp;

import android.graphics.Typeface;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.support.design.widget.NavigationView;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.Window;
import android.view.WindowManager;
import android.widget.TextView;
import android.widget.Toast;

import com.bhupendra.foodorderingapp.adapters.CategoryAdapter;
import com.bhupendra.foodorderingapp.data.CategoryInfo;
import com.bhupendra.foodorderingapp.utils.FirebaseConnector;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QuerySnapshot;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class MainActivity extends AppCompatActivity
        implements NavigationView.OnNavigationItemSelectedListener {

    SwipeRefreshLayout mRefreshLayout;
    RecyclerView mRecyclerLayout;
    CategoryAdapter categoryAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        Window w = getWindow();
        w.addFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);
        w.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION);

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mRecyclerLayout = findViewById(R.id.recyclerView);
        mRecyclerLayout.setLayoutManager(new GridLayoutManager(this, 2));
        mRefreshLayout = findViewById(R.id.swipeRefresh);

        refreshLayout();

        mRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                refreshLayout();
            }
        });

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        toolbar.setNavigationIcon(R.drawable.ic_mobile_menu);
        setSupportActionBar(toolbar);
        setTitle("");

        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        toggle.setDrawerIndicatorEnabled(false);
        toggle.setToolbarNavigationClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
                drawer.openDrawer(GravityCompat.START);
            }
        });
        drawer.addDrawerListener(toggle);
        toggle.syncState();

        NavigationView navigationView = (NavigationView) findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);
    }

    private void refreshLayout() {
        mRefreshLayout.setRefreshing(true);

        categoryAdapter = null;
        mRecyclerLayout.setAdapter(null);


        FirebaseFirestore db = FirebaseConnector.getFirebaseFirestore();
//
//        Map<String, Object> user = new HashMap<>();
//        user.put("first", "Ada");
//        user.put("last", "Lovelace");
//        user.put("born", 1815);
//
//        db.collection("users")
//                .add(user)
//                .addOnSuccessListener(new OnSuccessListener<DocumentReference>() {
//                    @Override
//                    public void onSuccess(DocumentReference documentReference) {
//                        Log.e("TestFirestore", documentReference.getId() + "Stored");
//                    }
//                })
//                .addOnFailureListener(new OnFailureListener() {
//                    @Override
//                    public void onFailure(@NonNull Exception e) {
//                        Log.e("TestFirestore", "Error adding document", e);
//                    }
//                });

        db.collection("menu")
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()) {

                            ArrayList<CategoryInfo> list = new ArrayList<>();

                            for (DocumentSnapshot document : task.getResult()) {
                                list.add(new CategoryInfo(document.getId(), (String) document.getData().get("image")));

//                                for (Object ob: objects) {
//                                    Map<String,Object> insideOb= (Map<String, Object>) ob;
//                                    list.add(new CategoryInfo((String) insideOb.get("name")));
//                                }
                            }

                            categoryAdapter = new CategoryAdapter(list, MainActivity.this);
                            mRecyclerLayout.setAdapter(categoryAdapter);
                            mRefreshLayout.setRefreshing(false);
                        } else {
                            Log.e("DataSnap", "Error getting documents.", task.getException());
                        }
                    }
                });

        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {

//                list.add(new CategoryInfo("HEllo"));
//                list.add(new CategoryInfo("Modi"));
//                list.add(new CategoryInfo("Chutiya"));
//                list.add(new CategoryInfo("Bina"));
//                list.add(new CategoryInfo("Lode"));
//                list.add(new CategoryInfo("Ka"));
//                list.add(new CategoryInfo("NBaal"));
//                list.add(new CategoryInfo("Hai"));
//                list.add(new CategoryInfo("Hijda"));
//                list.add(new CategoryInfo("MaakaLoda"));
//                list.add(new CategoryInfo("Chuda hua"));

            }
        }, 500);
    }

    @Override
    public void onBackPressed() {
        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }

    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        // Handle navigation view item clicks here.
        int id = item.getItemId();
//
//        if (id == R.id.nav_camera) {
//            // Handle the camera action
//        } else if (id == R.id.nav_gallery) {
//
//        } else if (id == R.id.nav_slideshow) {
//
//        } else if (id == R.id.nav_manage) {
//
//        } else if (id == R.id.nav_share) {
//
//        } else if (id == R.id.nav_send) {
//
//        }

        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        drawer.closeDrawer(GravityCompat.START);
        return true;
    }
}

File: app/src/main/java/com/adityagupta/foodorderingapp/activities/MainActivity.kt
package com.adityagupta.foodorderingapp.activities

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import com.google.android.material.navigation.NavigationView
import androidx.core.view.GravityCompat
import androidx.appcompat.app.ActionBarDrawerToggle
import android.view.MenuItem
import android.view.WindowManager
import com.adityagupta.foodorderingapp.adapters.CategoryAdapter
import com.adityagupta.foodorderingapp.data.CategoryInfo
import com.adityagupta.foodorderingapp.utils.FirebaseConnector
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.app_bar_main.*
import kotlinx.android.synthetic.main.content_main.*
import android.util.Log
import androidx.recyclerview.widget.GridLayoutManager
import com.adityagupta.foodorderingapp.R
import com.adityagupta.foodorderingapp.data.CategoryItem
import com.adityagupta.foodorderingapp.utils.Common


class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {

    private var categoryAdapter: CategoryAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        window.addFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS)
        window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION)

        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        recyclerView.layoutManager = GridLayoutManager(this, 2)

        refreshLayout()

        swipeRefresh.setOnRefreshListener { refreshLayout() }

        toolbar.setNavigationIcon(R.drawable.ic_mobile_menu)
        setSupportActionBar(toolbar)
        title = ""

        val toggle = ActionBarDrawerToggle(
                this, drawer_layout, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        toggle.isDrawerIndicatorEnabled = false
        toggle.setToolbarNavigationClickListener {
            drawer_layout.openDrawer(GravityCompat.START)
        }
        drawer_layout.addDrawerListener(toggle)
        toggle.syncState()

        nav_view.setNavigationItemSelectedListener(this)
    }

    private fun refreshLayout() {
        swipeRefresh.isRefreshing = true
        categoryAdapter = null
        recyclerView.adapter = null

        Common.categoryList = ArrayList()

        val db = FirebaseConnector.firebaseFirestore

        db.collection("menu")
                .get()
                .addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        for (document in task.result) {

                            var list: ArrayList<CategoryItem> = ArrayList()
                            for (item in  document.data["items"] as ArrayList<*>)
                            {
                                var map:HashMap<*,*> = (item as HashMap<*, *>)
                                var name :String = map.get("name") as String
                                var cat : String = map.get("category") as String
                                var price : Double = (map.get("price") as Any).toString().toDouble()

                                list.add(CategoryItem(name,cat,price))
                            }

                            Common.categoryList!!.add(CategoryInfo(document.id, document.data["image"] as String,list))

                            categoryAdapter = CategoryAdapter(Common.categoryList!!, this@MainActivity)
                            recyclerView.adapter = categoryAdapter
                            swipeRefresh.isRefreshing = false
                        }
                    } else {
                        Log.e("DataSnap", "Error getting documents.", task.exception)
                    }
                }
    }

    override fun onBackPressed() {
        if (drawer_layout.isDrawerOpen(GravityCompat.START)) {
            drawer_layout.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        drawer_layout.closeDrawer(GravityCompat.START)
        return true
    }

}
